lib/msal-node/src/client/ManagedIdentityApplication.ts

```diff
@@ -0,0 +1,137 @@
+ */
+export class ManagedIdentityApplication {
+    private managedIdentityId: ManagedIdentityId;
+    private config: ManagedIdentityNodeConfiguration;
+    private logger: Logger;
+    private cacheManager: CacheManager;
+    private networkManager: NetworkManager;
+
+    private identitySource:
+        | ServiceFabric
+        | AppService
+        | CloudShell
+        | AzureArc
+        | Imds;
+
+    constructor(
+        managedIdentityId: ManagedIdentityId,
+        configuration?: ManagedIdentityConfiguration
+    ) {
+        this.config = buildManagedIdentityConfiguration(configuration || {});
+        this.managedIdentityId = managedIdentityId;
+
+        this.logger = new Logger(
+            this.config.system.loggerOptions,
+            name,
+            version
+        );
+
+        this.cacheManager = new DefaultStorageClass(
+            this.managedIdentityId.id,
+            DEFAULT_CRYPTO_IMPLEMENTATION,
+            this.logger
+        );
+
+        const networkClient: INetworkModule = this.config.system.networkClient;
+        this.networkManager = new NetworkManager(
+            networkClient,
+            this.cacheManager
+        );
+
+        this.identitySource = this.selectManagedIdentitySource();
+    }
+
+    // TODO: implement this method
+    public async acquireToken(
+        request: ManagedIdentityRequest
+    ): Promise<AuthenticationResult | null> {
+        //
+    }
+
+    /**
+     * Tries to create a managed identity source for all sources
+     * @param request the managed identity request
+     * @returns the managed identity Source
+     */
+    // TODO: update return type
+    private selectManagedIdentitySource():
+        | ServiceFabric
+        | AppService
+        | CloudShell
+        | AzureArc
+        | Imds {
+        try {
+            return (
+                // parameters are based on DOTNET implementations
+                ServiceFabric.tryCreate(
+                    this.config,
+                    this.logger,
+                    this.cacheManager,
+                    this.networkManager,
+                    this.managedIdentityId
+                ) ||
+                AppService.tryCreate(
+                    this.config,
+                    this.logger,
+                    this.cacheManager,
+                    this.networkManager,
+                    this.managedIdentityId
+                ) ||
+                CloudShell.tryCreate(
+                    this.config,
+                    this.logger,
+                    this.cacheManager,
+                    this.networkManager,
+                    this.managedIdentityId
+                ) ||
+                AzureArc.tryCreate(
+                    this.config,
+                    this.logger,
+                    this.cacheManager,
+                    this.networkManager,
+                    this.managedIdentityId
+                ) ||
+                Imds.tryCreate(
```

Comment:
For IMDS, we directly call the constructor rather than try create the source. If none of the above sources are available IMDS will be used by default.

Reply:
I'm adding extra error checking in here. Maybe all the environment variables are malformed?